### PR TITLE
BIP 341: Specify Speedy Trial activation parameters

### PR DESCRIPTION
**Pull request by achow101**
Specifies activation parameters for BIP 341 using the Speedy Trial activation proposal discussed [on the bitcoin-dev mailing list](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-March/018594.html).

Testnet parameters were not discussed so I just chose block heights that corresponded to a start time of 2021/04/18 with 3 retarget periods of activation time and an additional 3 retarget periods for the minimum activation height.

Requires #1080 as it uses the `minimum_activation_height` specified there.

### CODE REVIEW

**Comment by michaelfolkson** [comment]
This includes (mistakenly I think) the commit making changes to BIP 8. This PR should only be making changes to BIP 341.

File: bip-0341.mediawiki [hunk 1]
```diff
@@ -294,7 +294,15 @@ Examples of preimage for sighashing for each of the sighash modes.
 
 == Deployment ==
 
-TODO
+There may be multiple deployment attempts of this BIP.
+
+The first deployment wil use BIP 8 with the name "taproot", using bit 2, and a lockinontimeout of false.
```
Comment (michaelfolkson) [inline bip-0341.mediawiki:299]: nit s/wil/will

File: bip-0341.mediawiki [hunk 2]
```diff
@@ -294,7 +294,15 @@ Examples of preimage for sighashing for each of the sighash modes.
...
+
+For Bitcoin mainnet, the BIP 8 startheight will be 681408, the timeoutheight will be 695520, the minimum_activation_height will be 709632, and the activation threshold will be 1815 out of 2016 blocks.
```
Comment (michaelfolkson) [inline bip-0341.mediawiki:301]: Perhaps follow "1815 out of 2016 blocks" with "(90 percent)" to make it clear a threshold of 90 percent is set.

File: bip-0341.mediawiki [hunk 3]
```diff
@@ -294,7 +294,15 @@ Examples of preimage for sighashing for each of the sighash modes.
...
+
+For Bitcoin testnet, the BIP 8 startheight will be 1945440, The timeoutheight will be 1951488, the minimum_activation_height will be 1957536, and the activation threshold will be 1512 out of 2016 blocks.
```
Comment (michaelfolkson) [inline bip-0341.mediawiki:303]: Perhaps follow "1512 out of 2016 blocks" with "(75 percent)" to make it clear a threshold of 75 percent is set.

**Review by michaelfolkson — COMMENTED** [review]
A few more nits.

File: bip-0341.mediawiki (hunk 1 above)
Comment (achow101) [reply bip-0341.mediawiki:299]: Done

File: bip-0341.mediawiki (hunk 2 above)
Comment (achow101) [reply bip-0341.mediawiki]: Done

File: bip-0341.mediawiki (hunk 3 above)
Comment (achow101) [reply bip-0341.mediawiki]: Done

**Comment by achow101** [comment]
> This includes (mistakenly I think) the commit making changes to BIP 8. This PR should only be making changes to BIP 341.

It is not mistaken. This PR requires BIP 8 to discuss the minimum activation height.

**Comment by Sjors** [comment]
Concept ACK. Can you add estimated dates next to the heights? I'd like to see https://github.com/bitcoin/bitcoin/pull/21392 merged and backported before committing to specific heights.

Ideally I'd also like to see another client implement the speedy trial.
* btcd is making some progress towards BIP 8 support, though no speedy trial yet: https://github.com/btcsuite/btcd/pull/1700. They also have no taproot implementation (only Schnorr WIP in https://github.com/btcsuite/btcd/pull/1667), but at least they would be able to detect a new softfork activated.
* bcoin has an old draft PR https://github.com/bcoin-org/bcoin/pull/948 to implement Taproot, but no BIP 8
* libbitcoin appears to have Taproot work in progress: https://github.com/libbitcoin/libbitcoin-system/issues/1177 but no BIP 8 yet

**Comment by achow101** [comment]
Added approximate dates

File: bip-0341.mediawiki [hunk 4]
```diff
@@ -294,7 +294,15 @@ Examples of preimage for sighashing for each of the sighash modes.
 
 == Deployment ==
 
-TODO
+There may be multiple deployment attempts of this BIP.
+
+The first deployment will use BIP 8 with the name "taproot", using bit 2, and a lockinontimeout of false.
+
+For Bitcoin mainnet, the BIP 8 startheight will be 681408 (approximately May 1st 2021), the timeoutheight will be 695520 (approximately August 6th 2021), the minimum_activation_height will be 709632 (approximately Novermber 12th 2021), and the activation threshold will be 1815 out of 2016 blocks (90%).
```
Comment (flack) [inline bip-0341.mediawiki:301]:
```suggestion
For Bitcoin mainnet, the BIP 8 startheight will be 681408 (approximately May 1st 2021), the timeoutheight will be 695520 (approximately August 6th 2021), the minimum_activation_height will be 709632 (approximately November 12th 2021), and the activation threshold will be 1815 out of 2016 blocks (90%).
```

Comment (achow101) [reply bip-0341.mediawiki]: Oops, fixed.

**Comment by benthecarman** [comment]
ACK 3704017f7ba4e0c152c805d6af9969be48d30cdd

File: bip-0341.mediawiki [hunk 5]
```diff
@@ -294,7 +294,15 @@ Examples of preimage for sighashing for each of the sighash modes.
...
 == Deployment ==
 
-TODO
+There may be multiple deployment attempts of this BIP.
+
+The first deployment will use BIP 8 with the name "taproot", using bit 2, and a lockinontimeout of false.
+
+For Bitcoin mainnet, the BIP 8 startheight will be 681408 (approximately May 1st 2021), the timeoutheight will be 695520 (approximately August 6th 2021), the minimum_activation_height will be 709632 (approximately November 12th 2021), and the activation threshold will be 1815 out of 2016 blocks (90%).
+
+For Bitcoin testnet, the BIP 8 startheight will be 1945440 (approximately April 20th 2021), The timeoutheight will be 1951488 (approximately June 1st 2021), the minimum_activation_height will be 1957536 (approximately July 13th 2021), and the activation threshold will be 1512 out of 2016 blocks (75%).
```
Comment (ajtowns) [inline bip-0341.mediawiki:303]: The entire testnet range from block [1945440](https://blockstream.info/testnet/block/0000000000003531d7cc792e8fc02bfbbb701931f6a06751781a80329833c3d3) to [1957536](https://blockstream.info/testnet/block/0000000000003d926cb25f3840d58b6f1536c91bae32f6b145699bb8443b64f5) was mined today, with no signalling on bit 2 (actually no signalling except on the bip 320 reserved bits).

Probably doesn't make sense to use future heights for testnet.

Comment (achow101) [reply bip-0341.mediawiki:303]: Ah that's unfortunate. Maybe it should just be always active? Gonna try seeing if I can sync a testnet node with that.

Comment (ajtowns) [reply bip-0341.mediawiki:303]: I was thinking just add activation rules for testnet after it's active on mainnet, maybe even after it's buried on mainnet?

Comment (achow101) [reply bip-0341.mediawiki:303]: ngl feels like we should abandon testnet3 really.

Comment (michaelfolkson) [reply bip-0341.mediawiki:303]: Agree with AJ's suggestion to activate on testnet once active on mainnet given Taproot is available for testing on signet. The fewer heights we have to keep adjusting the better.

Comment (achow101) [reply bip-0341.mediawiki]: Dropped these testnet params and marked it as TBD.

**Comment by ajtowns** [comment]
This misses activating BIP342 -- suggest rebasing on #1091 to fix that.

**Comment by achow101** [comment]
Rebased onto #1091

File: bip-0341.mediawiki [hunk 6]
```diff
@@ -294,7 +294,17 @@ Examples of preimage for sighashing for each of the sighash modes.
...
+
+For Bitcoin signet, these BIPs are always active.
+
+There may be multiple deployment attempts of these BIPs.
+
+The first deployment will use BIP 8 with the name "taproot", using bit 2, and a lockinontimeout of false.
+
+For Bitcoin mainnet, the BIP 8 startheight will be 681408 (approximately May 1st 2021), the timeoutheight will be 695520 (approximately August 6th 2021), the minimum_activation_height will be 709632 (approximately November 12th 2021), and the activation threshold will be 1815 out of 2016 blocks (90%).
+
+For Bitcoin testnet, TBD.
```
Comment (Sjors) [inline bip-0341.mediawiki:307]: Maybe elaborate here:

> For Bitcoin testnet, TBD. Due to the difficulty reset, the pace of testnet block generation is unpredictable. This makes height based activation impractical.

Comment (michaelfolkson) [reply bip-0341.mediawiki:307]: Or alternatively:

"For Bitcoin testnet, Taproot will be activated if and when Taproot is active on mainnet. Taproot is already active on signet for testing and experimentation purposes."

Comment (achow101) [reply bip-0341.mediawiki]: Added these proposed heights.

**Comment by Rspigler** [comment]
Should this be closed now?

**Pull request closed without merging**